### PR TITLE
1.20.5 update

### PR DESCRIPTION
**Pull request by booky10**
~~Still WIP, 1.20.5 isn't released yet.~~ Now it is, but this PR still needs more testing.

### Todo

- [x] Create new packet wrappers
    - [x] WrapperConfigServerResetChat
    - [x] WrapperConfigServerSelectKnownPacks
    - [x] WrapperPlayServerDebugSample
    - [x] WrapperPlayServerProjectilePower
    - [x] WrapperConfigClientSelectKnownPacks
    - [x] WrapperPlayClientChatCommandUnsigned
    - [x] WrapperPlayClientDebugSampleSubscription
- [x] Update existing packet wrappers
- [x] Update block entity types
- [x] Add mappings for command argument parsers
- [x] Add mappings for potion types
- [x] New component system
    - [x] New ItemStack serialization
    - [x] Component Type Registry
    - [x] Implement serializer for components
    - [x] Add default components for all ItemTypes
    - [x] Make component contents mutable
    - [x] Fix/Add equality checks for all components
    - [x] Fix ArmorTrim Pattern/Material serialization
- [x] Add mappings to sound registry for <=1.20.5
- [x] Fix entity sound packets not supporting direct serialization

---

Some stuff worth mentioning:
- This introduces mappings for attributes, as new ones were added
- This introduces mappings for block entity types, as these change frequently and there didn't seem to exist any yet
- This introduces mappings for command argument parsers, as these changed throughout the versions and were just always for the latest version
- The `sweeping` enchantment was renamed to `sweeping_edge`, like done in 24w03a
- The `scute` item was renamed to `turtle_scute`
- The tag field names have been adjusted to be more consistent with the vanilla tag names
- ItemStack serialization was completely rewritten for the new component system:
    - ItemTypes now have some default components set (not sent over network)
    - ItemStacks can override these default components (remove or replace) (sent over network)
      Every component has their own serializer which writes bytes instead of nbt (like previously)
- Registry data will need to be tracked on login in some point in the future to support datapacks (and bukkit...) changing customizable vanilla registries (like chat_type, trim_pattern, trim_material, etc.)

### CODE REVIEW

**Comment by SamB440** [comment]
Spigot 1.20.5
```
[17:51:32] [Netty Server IO #1/WARN]: Failed to initialize a channel. Closing: [id: 0xfcb31cef, L:/127.0.0.1:25565 - R:/127.0.0.1:54457]
java.lang.IllegalStateException: PacketEvents failed to add a decoder to the netty pipeline. Pipeline handlers: [pe-connection-initializer-packetevents, timeout, legacy_query, splitter, FlowControlHandler#0, pe-decoder-packetevents, decoder, prepender, outbound_config, hackfix, packet_handler, DefaultChannelPipeline$TailContext#0]
        at io.github.retrooper.packetevents.injector.connection.ServerConnectionInitializer.relocateHandlers(ServerConnectionInitializer.java:119) ~[?:?]
        at io.github.retrooper.packetevents.injector.connection.ServerConnectionInitializer.initChannel(ServerConnectionInitializer.java:73) ~[?:?]
        at io.github.retrooper.packetevents.injector.connection.PreChannelInitializer_v1_12.channelRegistered(PreChannelInitializer_v1_12.java:34) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRegistered(AbstractChannelHandlerContext.java:176) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRegistered(AbstractChannelHandlerContext.java:152) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRegistered(AbstractChannelHandlerContext.java:145) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRegistered(DefaultChannelPipeline.java:1383) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRegistered(AbstractChannelHandlerContext.java:172) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRegistered(AbstractChannelHandlerContext.java:152) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRegistered(DefaultChannelPipeline.java:815) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:517) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:429) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174) ~[netty-common-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167) ~[netty-common-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[netty-common-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.97.Final.jar:4.1.97.Final]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.util.NoSuchElementException: encoder
        at io.netty.channel.DefaultChannelPipeline.getContextOrDie(DefaultChannelPipeline.java:1073) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:248) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:237) ~[netty-transport-4.1.97.Final.jar:4.1.97.Final]
        at io.github.retrooper.packetevents.injector.connection.ServerConnectionInitializer.relocateHandlers(ServerConnectionInitializer.java:116) ~[?:?]
        ... 19 more
```

File: api/src/main/java/com/github/retrooper/packetevents/protocol/attribute/Attributes.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    // Added in 1.20.2
+    public static final Attribute GENERIC_MAX_ABSORPTION = define("generic.max_absorption");
+
+    // added in 1.20.5
+    public static final Attribute PLAYER_BLOCK_BREAK_SPEED = define("player.block_break_speed");
+    public static final Attribute PLAYER_BLOCK_INTERACTION_RANGE = define("player.block_interaction_range");
+    public static final Attribute PLAYER_ENTITY_INTERACTION_RANGE = define("player.entity_interaction_range");
+    public static final Attribute GENERIC_FALL_DAMAGE_MULTIPLIER = define("generic.fall_damage_multiplier");
+    public static final Attribute GENERIC_GRAVITY = define("generic.scale");
```
Comment (SamB440) [inline api/src/main/java/com/github/retrooper/packetevents/protocol/attribute/Attributes.java:108]:
```suggestion
    public static final Attribute GENERIC_GRAVITY = define("generic.gravity");
```

**Review by SamB440 — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/github/retrooper/packetevents/protocol/item/type/ItemTypes.java [hunk 2]
```diff
@@ -104,68 +153,68 @@ public class ItemTypes {
...
+    public static final ItemType GOLDEN_AXE = builder("golden_axe").setMaxAmount(1).setMaxDurability(32).setAttributes(ItemAttribute.GOLD_TIER, ItemAttribute.AXE).setComponent(DAMAGE, 0).setComponent(ATTRIBUTE_MODIFIERS, new ItemAttributeModifiers(Arrays.asList(new ItemAttributeModifiers.ModifierEntry(Attributes.GENERIC_ATTACK_DAMAGE, new ItemAttributeModifiers.Modifier(UUID.fromString("cb3f55d3-645c-4f38-a497-9c13a33db5cf"), "Tool modifier", 6.0d, AttributeOperation.ADDITION), ItemAttributeModifiers.EquipmentSlotGroup.MAINHAND), new ItemAttributeModifiers.ModifierEntry(Attributes.GENERIC_ATTACK_SPEED, new ItemAttributeModifiers.Modifier(UUID.fromString("fa233e1c-4180-4865-b01b-bcce9785aca3"), "Tool modifier", -3.0d, AttributeOperation.ADDITION), ItemAttributeModifiers.EquipmentSlotGroup.MAINHAND)), true)).setComponent(TOOL, new ItemTool(Arrays.asList(new ItemTool.Rule(new MappedEntitySet<>(ResourceLocation.minecraft("incorrect_for_gold_tool")), null, false), new ItemTool.Rule(new MappedEntitySet<>(ResourceLocation.minecraft("mineable/axe")), 12.0f, true)), 1.0f, 1)).build();
```
Comment (SamB440) [inline api/src/main/java/com/github/retrooper/packetevents/protocol/item/type/ItemTypes.java:166]: Oh this stuff seems a bit cursed with how long it is, I think we should at least make the UUIDS a static constant as they seem to be referenced in the other axes too

**Comment by retrooper** [comment]
For the next PacketEvents release, we planned to merge both this PR and https://github.com/retrooper/packetevents/pull/648/files#diff-d793ba1bcd8a7285f509a33d48e278d16a8d8318d2189761a09f41bb6fa438a4 
We need to make sure we can fix possible conflicts.

**Comment by AbhigyaKrishna** [comment]
Thank you for the pr!!

**Pull request merged**